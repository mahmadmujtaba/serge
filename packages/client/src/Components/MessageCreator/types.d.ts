interface Props {
  schema: any,
  curChannel: string,
  privateMessage: string
  onMessageSend?: (e: any) => void
  onCancel?: (e: any) => void
}
export default Props
