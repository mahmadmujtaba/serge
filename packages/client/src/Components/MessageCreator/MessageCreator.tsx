import React, { useState, createRef, useEffect } from 'react'
import { faUserSecret } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { saveMessage } from '../../ActionsAndReducers/playerUi/playerUi_ActionCreators'
import { usePlayerUiState } from '../../Store/PlayerUi'
import { Editor, MessageDetails } from '@serge/custom-types'
import Props from './types'

// @ts-ignore
import JSONEditor from '@json-editor/json-editor'

import '@serge/themes/App.scss'

const MessageCreator: React.FC<Props> = (props) => {

  const [editor, setEditor] = useState<Editor | null>(null)
  const editorPreviewRef = createRef<HTMLDivElement>()
  const privateMessageRef = createRef<HTMLTextAreaElement>()
  const [selectedSchema, setSelectedSchema] = useState<any>(props.schema)
  const state = usePlayerUiState()

  const sendMessage = (): void => {
    const curForce = state.allForces.find((force) => force.uniqid === state.selectedForce)
    if (!curForce) return

    const details: MessageDetails = {
      channel: props.curChannel,
      from: {
        force: curForce.name,
        forceColor: state.forceColor,
        role: state.selectedRole,
        icon: curForce.icon,
      },
      messageType: selectedSchema.title,
      timestamp: new Date().toISOString(),
    }

    if (props.privateMessage && privateMessageRef.current) {
      details.privateMessage = privateMessageRef.current.value
      privateMessageRef.current.value = ''
    }

    if (!editor || editor.getValue().content === '') return

    saveMessage(state.currentWargame, details, editor.getValue())()
    editor.destroy()
    setEditor(null)
    createEditor(selectedSchema)
  }

  useEffect(() => {
    if (props.schema && (!selectedSchema || selectedSchema.title !== props.schema.title)) {
      if(editor) editor.destroy()
      setEditor(null)
      console.log(props.schema, 'props.schema');
      setSelectedSchema(props.schema)
    }

    if (props.schema && props.schema.type) {
      if (editor) return
      createEditor(props.schema)
    }
  }, [props])

  const createEditor = (schema: any) => {
    setEditor(new JSONEditor(editorPreviewRef.current, {
      schema,
      theme: 'bootstrap4',
      disable_collapse: true,
      disable_edit_json: true,
      disable_properties: true,
    }))
  }

  return (
    <>
      <div className='form-group message-creator' ref={editorPreviewRef}/>
      {props.privateMessage && <div className='flex-content form-group'>
        <label htmlFor='' className='material-label' id='private-message-input-label'>
          <FontAwesomeIcon size='2x' icon={faUserSecret}/>Private message
        </label>
        <textarea id='private-message-input' className='form-control' ref={privateMessageRef} />
      </div>}
      <div className='form-group'>
        <button name='send' className='btn btn-action btn-action--form' onClick={sendMessage}>
          <span>Send Message</span>
        </button>
      </div>
    </>
  )
}

export default MessageCreator
