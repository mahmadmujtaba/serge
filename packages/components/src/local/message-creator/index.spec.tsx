/* global it expect */

import React from 'react'
import renderer from 'react-test-renderer'

import MessageCreator from './index'

it('MessageCreator renders correctly', () => {
  const tree = renderer
    .create(<MessageCreator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});