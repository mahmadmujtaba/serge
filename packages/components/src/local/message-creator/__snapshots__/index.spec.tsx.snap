// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageCreator renders correctly 1`] = `
<div>
  <section
    className="input-container"
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
        onClick={[Function]}
      >
        <textarea
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={2}
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Send
      </span>
    </button>
  </section>
</div>
`;
